flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,218 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Abstraction used by the {@link Dispatcher} to manage JobManagers, in
+ * particular during initialization.
+ * While a job is initializing, the JobMasterGateway is not available. A small subset
+ * of the methods of the JobMasterGateway necessary during initialization are provided
+ * by this class (job details, cancel).
+ */
+public final class DispatcherJob implements AutoCloseableAsync {
+
+	private final Logger log = LoggerFactory.getLogger(getClass());
+
+	private final CompletableFuture<JobManagerRunner> jobManagerRunnerFuture;
+	private final CompletableFuture<ArchivedExecutionGraph> jobResultFuture;
+
+	private final Object lock = new Object();
+
+	private final JobGraph jobGraph;
+	private final long initializationTimestamp;
+
+	// internal field to track job status during initialization. Is not updated anymore after
+	// job is initialized, cancelled or failed.
+	@GuardedBy("lock")
+	private JobStatus jobStatus = JobStatus.INITIALIZING;
+
+	@Nullable
+	@GuardedBy("lock")
+	private CompletableFuture<Acknowledge> cancellationFuture = null;
+
+	private enum SubmissionType {
+		INITIAL, RECOVERY
+	}
+
+	static DispatcherJob createForSubmission(CompletableFuture<JobManagerRunner> jobManagerRunnerFuture, JobGraph jobGraph, long initializationTimestamp) {
+		return new DispatcherJob(jobManagerRunnerFuture, jobGraph, initializationTimestamp, SubmissionType.INITIAL);
+	}
+
+	static DispatcherJob createForRecovery(CompletableFuture<JobManagerRunner> jobManagerRunnerFuture, JobGraph jobGraph, long initializationTimestamp) {
+		return new DispatcherJob(jobManagerRunnerFuture, jobGraph, initializationTimestamp, SubmissionType.RECOVERY);
+	}
+
+	private DispatcherJob(CompletableFuture<JobManagerRunner> jobManagerRunnerFuture,
+		JobGraph jobGraph, long initializationTimestamp, SubmissionType submissionType) {
+		this.jobManagerRunnerFuture = jobManagerRunnerFuture;
+		this.jobGraph = jobGraph;
+		this.initializationTimestamp = initializationTimestamp;
+		this.jobResultFuture = new CompletableFuture<>();
+
+		FutureUtils.assertNoException(this.jobManagerRunnerFuture.handle((jobManagerRunner, throwable) -> {
+			// JM has been initialized, or the initialization failed
+			synchronized (lock) {
+				if (throwable == null) {
+					if (cancellationFuture != null) {
+						Preconditions.checkState(jobStatus == JobStatus.CANCELLING);
+						log.warn(
+							"JobManager initialization has been cancelled for {}. Stopping JobManager.",
+							jobGraph.getJobID());
+
+						CompletableFuture<Acknowledge> cancelJobFuture = jobManagerRunner.getJobMasterGateway().thenCompose(
+							gw -> gw.cancel(Time.seconds(60))); // TODO set better timeout
+						cancelJobFuture.whenComplete((ack, cancelThrowable) -> {
+							jobStatus = JobStatus.CANCELED;
```

Comment:
`jobStatus` is not guarded by `lock`.

Reply:
After `gw.cancel` completes it is not guaranteed that the Job has reached the state `JobStatus.CANCELED`.